src/resource.rs

```diff
@@ -2,35 +2,291 @@
+        if args.len() != self.args.len() {
+            return Err(SignatureError::TypeArgMismatch(
+                "Node provided wrong number of args".to_string(),
+            ));
+        }
+        for (a, p) in args.iter().zip(self.args.iter()) {
+            check_arg(a, p).map_err(SignatureError::TypeArgMismatch)?;
+        }
+        let (ins, outs, res) = match &self.signature_func {
+            SignatureFunc::FromYAML { .. } => {
+                // Sig should be computed solely from inputs + outputs + args.
+                // Resources used should be at least the resource containing this OpDef.
+                // (TODO Consider - should we identify that _in_ the OpDef?)
+                todo!()
+            }
+            SignatureFunc::CustomFunc(bf) => bf.compute_signature(&self.name, args, &self.misc)?,
+        };
+        let mut sig = Signature::new_df(ins, outs);
+        sig.input_resources = resources_in.clone();
+        sig.output_resources = res.union(resources_in); // Pass input requirements through
+        Ok(sig)
+    }
 
-use crate::ops::{OpDef, OpaqueOp};
-use crate::types::custom::CustomType;
+    /// Optional description of the ports in the signature.
+    pub fn signature_desc(&self, args: &Vec<TypeArgValue>) -> SignatureDescription {
+        match &self.signature_func {
+            SignatureFunc::FromYAML { .. } => {
+                todo!()
+            }
+            SignatureFunc::CustomFunc(bf) => bf.describe_signature(&self.name, args, &self.misc),
+        }
+    }
+
+    /// Fallibly returns a Hugr that may replace an instance of this OpDef
+    /// given a set of available resources that may be used in the Hugr.
+    pub fn try_lower(
+        &self,
+        args: &Vec<TypeArgValue>,
+        available_resources: &ResourceSet,
+    ) -> Option<Hugr> {
+        match &self.lower_func {
+            LowerFunc::None => None,
+            LowerFunc::FixedHugr(req_res, h) => {
+                if available_resources.is_superset(req_res) {
+                    Some(h.clone())
+                } else {
+                    None
+                }
+            }
+            LowerFunc::CustomFunc(f) => {
+                f.try_lower(&self.name, args, &self.misc, available_resources)
+            }
+        }
+    }
+}
 
 /// A unique identifier for a resource.
 ///
 /// The actual [`Resource`] is stored externally.
 pub type ResourceId = SmolStr;
 
 /// A resource is a set of capabilities required to execute a graph.
-#[derive(Clone, Debug, Default, serde::Serialize, serde::Deserialize)]
+#[derive(Debug, Default, serde::Serialize, serde::Deserialize)]
 pub struct Resource {
     /// Unique identifier for the resource.
-    pub name: ResourceId,
-    /// Set of resource dependencies required by this resource.
-    pub resource_reqs: ResourceSet,
+    name: ResourceId,
+    // Set of resource dependencies required by this resource.
+    // TODO I haven't seen where these are required yet. If they are,
```

Comment:
What about the case where someone wants to define a resource in terms of types provided by another resource?

This doesn't necessarily require any knowledge of the implementation of the other resource - one could have an operation on types from another resource which is defined in terms of that resources operations

Reply:
Yes, indeed. I'm not quite sure what this should look like. It may fall under the heading of "namespacing", which is one of the unresolved issues in this PR. Good point, thanks :)

Reply:
I've kept this in as an "upper bound" of all signatures that *might* be returned by compute_signature functions, which can't be determined by inspection.

However, I'm not entirely sure how it's used. If I declare an operation that takes a MyType<Int,6> without the resource declaring the type MyType being around (even in YAML!), seems like I don't lose anything - it's just an early-warning that I the end-user-programmer *may* not be able to get very far. But the HUGR is still capable of passing such values around as serialized blobs, if I have operations that both create and consume those values. Really, the only information added by the "declaration" of the custom type is validity checking of the type arguments?!